File: client/Maui/Services/ApiService.cs
namespace Blazor.Services;

using System.Net.Http.Json;

public class ApiService
{
    private readonly HttpClient _httpClient;

    public ApiService(HttpClient httpClient)
    {
        _httpClient = httpClient;
        _httpClient.BaseAddress = new Uri("http://localhost:5000/api/");
    }

    public async Task<T> Get<T>(string controller, string action)
    {
        var res = await _httpClient.GetFromJsonAsync<T>($"{controller}/{action}");
        return res;
    }

    public async Task<IEnumerable<T>> GetAll<T>(string controller)
    {
        var res = await _httpClient.GetFromJsonAsync<IEnumerable<T>>(controller);
        return res;
    }

    public async Task<T> GetSingle<T>(string controller, string id)
    {
        var res = await _httpClient.GetFromJsonAsync<T>($"{controller}/{id}");
        return res;
    }

    public async Task<T> Post<T>(string controller, T itemToUpdate)
    {
        var res = await _httpClient.PostAsJsonAsync(controller, itemToUpdate);
        return default;
    }

    public async Task<T> Update<T>(string controller, string id, T itemToUpdate)
    {
        var res = await _httpClient.PutAsJsonAsync($"{controller}/{id}", itemToUpdate);
        return default;
    }

    public async Task<T> Delete<T>(string controller, string id)
    {
        var res = await _httpClient.DeleteAsync($"{controller}/{id}");
        return default;
    }
}

File: client/Angular/src/app/svc/api.service.ts
import { HttpClient, HttpEvent, HttpHandler, HttpInterceptor, HttpRequest } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { ElectronService } from 'ngx-electron';

@Injectable()
export class ApiService {

  private actionUrl: string;
  private _isElectronApp: boolean;

  constructor(electron: ElectronService, private http: HttpClient) {
    this._isElectronApp = electron.isElectronApp;
  }

  public getAll<T>(controller: string): Observable<T> {
    return this.http.get<T>(this.generateUrl(controller));
  }

  public getSingle<T>(controller: string, id: string): Observable<T> {
    return this.http.get<T>(this.generateUrl(controller) + id);
  }

  public post<T>(controller: string, itemToUpdate: any): Observable<T> {
    return this.http
      .post(this.generateUrl(controller), itemToUpdate) as Observable<T>;
  }

  public update<T>(controller: string, id: string, itemToUpdate: any): Observable<T> {
    return this.http
      .put<T>(this.generateUrl(controller) + id, itemToUpdate);
  }

  public delete<T>(controller: string, id: string): Observable<T> {
    return this.http.delete<T>(this.generateUrl(controller) + id);
  }

  public generateUrl(controller: string) {
    let apiUrl = `api/${controller}/`;
    if (this._isElectronApp) {
      apiUrl = `http://localhost:5000/${apiUrl}`;
    }
    return apiUrl;
  }
}


@Injectable()
export class CustomInterceptor implements HttpInterceptor {

  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    if (!req.headers.has('Content-Type')) {
      req = req.clone({ headers: req.headers.set('Content-Type', 'application/json') });
    }

    req = req.clone({ headers: req.headers.set('Accept', 'application/json') });
    console.log(JSON.stringify(req.headers));
    return next.handle(req);
  }
}
